Controllers/CHumidityController.cpp

```diff
@@ -0,0 +1,261 @@
+/*
+ * CHumidityController.cpp
+ *
+ *  Created on: Aug. 11, 2022
+ *      Author: samk
+ */
+
+#include "CHumidityController.h"
+#include <stdio.h>
+#include "CBME280.h"
+#include "ICommand.h"
+#include "IHardwareMap.h"
+
+#define MIN_TEMPERATURE  10
+#define MAX_TEMPERATURE  50
+#define MIN_POWER        0
+#define MAX_POWER        100
+#define DISABLE_OVERRIDE -1
+#define DISABLE_TARGET   0
+#define MIN_HUMIDITY     85
+#define MAX_HUMIDITY     95
+
+CHumidityController::CHumidityController(IHardwareMap *p_hardware,
+                                         etl::string<MAX_STRING_SIZE> name,
+                                         uint32_t run_period_ms,
+                                         SPI_HandleTypeDef *p_spi,
+                                         GPIO_TypeDef *p_slave_select_port,
+                                         uint16_t slave_select_pin)
+    : CController(name, run_period_ms),
+      mp_hw(p_hardware)
+{
+    // TODO Auto-generated constructor stub
+    reset();
+
+    CBME280 hum_temp_reader(p_spi, p_slave_select_port, slave_select_pin);
+    hum_sensor = &hum_temp_reader;
+}
+
+/**
+ * @brief Run all repeated humidity control activities.
+ *
+ */
+void CHumidityController::run()
+{
+    float actual_humidity;
+    for (int i = 0; i < CHANNEL_NUMBER; i++)
+    {
+        float power = 0;
+        actual_humidity = hum_sensor->getHumidity();
+        float actual_temperature = hum_sensor->getTemperature();
+        if (m_power_override[i] == DISABLE_OVERRIDE &&
+            target_humidity != DISABLE_TARGET)
+        {
+            if (actual_humidity != target_humidity)
+            {
+                if (actual_humidity < target_humidity)
+                {
+                    power = m_control_loop[i].run(actual_temperature,
+                                                  MIN_TEMPERATURE);
+                }
+                else
+                {
+                    power = m_control_loop[i].run(actual_temperature,
+                                                  MAX_TEMPERATURE);
+                }
+            }
+            else
+            {
+                break;
+            }
+        }
+        else
+        {
+            power = m_power_override[i];
+        }
+        mp_hw->setHardPwmOutput(power, i);
+    }
+    while (actual_humidity != target_humidity)
+    {
+        run();
+    }
+}
+
+bool CHumidityController::newCommand(ICommand *p_command,
+                                     IComChannel *p_comchannel)
+{
+    bool b_command_recognised = false;
+    ICommand::command_error_code_t result = ICommand::COMMAND_OK;
+    /**
+     * Query about current state of all active channels.
```

Comment:
One of the things you will need to learn is where to comment. One habit that is worth developing is avoiding comments. If you feel like a piece of code requires comment, then first thing you should consider is whether you should refactor it to make code very clear and not require comments.
This type of comment is superfluous. Your code is very clean and easy to read, so cluttering it with comments that do not really add new information is detrimental. 
Generally, comments should be placed where complex piece of logic is happening, e.g. you're doing some numerical approximation using your own algorithm. Or you have implemented another algorithm in which case brief description is in order and then a reference to a book or a website where algorithm is described in details. Another possibility for comments are in places where you have optimised the `if... else` construct for execution speed and this reduced readability, but you really want speed improvement, so explain what you've done. 
This will come with experience, so don't worry about it too much now, but try to follow the principle.